# Flight Operations Data Analysis:
## Objective
The objective of this project is to perform a complete data analysis of a Flight Operations dataset using Python, NumPy, and Pandas. The analysis includes understanding the dataset, inspecting its structure and statistics, selecting and filtering data, sorting records, grouping and aggregating data, applying data transformations, performing datetime operations, and using merge, join, and concatenate operations to derive meaningful insights and identify patterns in flight operations.

In [242]:
import numpy as np
import pandas as pd

## 1. Loading the Dataset
The Flight Operations dataset is stored in CSV format. Pandas `read_csv()` is used to load the CSV file into a DataFrame.

In [243]:
df = pd.read_csv("Day10_Flight_Operations_Dataset.csv")

In [244]:
df

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,0.0,On Time,Fog,Travel Agency,16,Vegan,5
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737,Economy,67,220,4713,5.0,On Time,Rain,Travel Agency,19,No Meal,2
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,35.0,Delayed,Storm,Travel Agency,18,No Meal,2
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,20.0,Delayed,Cloudy,Airline Website,25,No Meal,5
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
175,FL0176,2026-04-10,IndiGo,Ahmedabad,Bengaluru,Airbus A320,Economy,48,189,4298,0.0,On Time,Storm,Mobile App,11,Vegetarian,5
176,FL0177,2026-02-17,Vistara,Chennai,Mumbai,Airbus A320,Business,124,180,17306,20.0,Delayed,Cloudy,Mobile App,8,No Meal,5
177,FL0178,2026-04-30,IndiGo,Delhi,Mumbai,Airbus A319,Economy,112,186,4490,110.0,Delayed,Fog,Online Travel Portal,16,Vegetarian,3
178,FL0179,2026-05-08,Vistara,Bengaluru,Kochi,Airbus A320,Economy,134,189,5413,20.0,Delayed,Rain,Online Travel Portal,9,No Meal,2


## 2. Initial Data Inspection
Before analyzing the data, we inspect the first and last records, dataset dimensions, column names, data types, and statistical summary.

In [245]:
df.head()

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,0.0,On Time,Fog,Travel Agency,16,Vegan,5
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737,Economy,67,220,4713,5.0,On Time,Rain,Travel Agency,19,No Meal,2
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,35.0,Delayed,Storm,Travel Agency,18,No Meal,2
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,20.0,Delayed,Cloudy,Airline Website,25,No Meal,5


In [246]:
df.tail()

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
175,FL0176,2026-04-10,IndiGo,Ahmedabad,Bengaluru,Airbus A320,Economy,48,189,4298,0.0,On Time,Storm,Mobile App,11,Vegetarian,5
176,FL0177,2026-02-17,Vistara,Chennai,Mumbai,Airbus A320,Business,124,180,17306,20.0,Delayed,Cloudy,Mobile App,8,No Meal,5
177,FL0178,2026-04-30,IndiGo,Delhi,Mumbai,Airbus A319,Economy,112,186,4490,110.0,Delayed,Fog,Online Travel Portal,16,Vegetarian,3
178,FL0179,2026-05-08,Vistara,Bengaluru,Kochi,Airbus A320,Economy,134,189,5413,20.0,Delayed,Rain,Online Travel Portal,9,No Meal,2
179,FL0180,2026-01-22,Akasa Air,Delhi,Mumbai,Airbus A321,Economy,144,210,3767,75.0,Delayed,Fog,Online Travel Portal,17,No Meal,5


In [247]:
df.sample(5)

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
113,FL0114,2026-06-27,Air India Express,Srinagar,Delhi,Airbus A320,Economy,79,180,6268,20.0,Delayed,Clear,Mobile App,21,Vegan,5
126,FL0127,2026-01-21,IndiGo,Delhi,Chennai,Airbus A321,Economy,184,180,4960,5.0,On Time,Fog,Mobile App,10,Vegetarian,5
39,FL0040,2026-03-21,Vistara,Kochi,Delhi,Airbus A319,Economy,178,220,5361,12.0,On Time,Cloudy,Mobile App,20,Vegan,4
123,FL0124,2026-01-06,Vistara,Delhi,Chennai,Airbus A319,Economy,138,210,5399,0.0,On Time,Cloudy,Mobile App,11,No Meal,3
55,FL0056,2026-05-02,Air India Express,Pune,Delhi,Airbus A320,Economy,180,180,6397,5.0,On Time,Fog,Mobile App,14,Non-Vegetarian,4


In [248]:
df.shape


(180, 17)

In [249]:
df.columns

Index(['Flight_ID', 'Flight_Date', 'Airline', 'Origin', 'Destination',
       'Aircraft', 'Travel_Class', 'Passengers', 'Seat_Capacity',
       'Average_Ticket_Price', 'Delay_Minutes', 'Flight_Status', 'Weather',
       'Booking_Channel', 'Avg_Baggage_Kg', 'Meal_Preference',
       'Passenger_Satisfaction'],
      dtype='str')

In [250]:
df.dtypes

Flight_ID                     str
Flight_Date                   str
Airline                       str
Origin                        str
Destination                   str
Aircraft                      str
Travel_Class                  str
Passengers                  int64
Seat_Capacity               int64
Average_Ticket_Price        int64
Delay_Minutes             float64
Flight_Status                 str
Weather                       str
Booking_Channel               str
Avg_Baggage_Kg              int64
Meal_Preference               str
Passenger_Satisfaction      int64
dtype: object

In [251]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 180 entries, 0 to 179
Data columns (total 17 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Flight_ID               180 non-null    str    
 1   Flight_Date             180 non-null    str    
 2   Airline                 180 non-null    str    
 3   Origin                  180 non-null    str    
 4   Destination             180 non-null    str    
 5   Aircraft                180 non-null    str    
 6   Travel_Class            180 non-null    str    
 7   Passengers              180 non-null    int64  
 8   Seat_Capacity           180 non-null    int64  
 9   Average_Ticket_Price    180 non-null    int64  
 10  Delay_Minutes           173 non-null    float64
 11  Flight_Status           180 non-null    str    
 12  Weather                 180 non-null    str    
 13  Booking_Channel         180 non-null    str    
 14  Avg_Baggage_Kg          180 non-null    int64  
 15  

In [252]:
df.describe()

,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Avg_Baggage_Kg,Passenger_Satisfaction
count,180.000000,180.000000,180.000000,173.000000,180.000000,180.000000
mean,122.772222,197.238889,6111.677778,19.335260,15.150000,3.494444
std,45.635651,15.408176,3615.220040,29.029712,5.705506,1.169908
min,45.000000,180.000000,1949.000000,0.000000,5.000000,2.000000
25%,78.750000,186.000000,3815.500000,0.000000,10.000000,2.000000
50%,125.000000,189.000000,5380.000000,5.000000,15.500000,3.000000
75%,162.250000,210.000000,6790.000000,20.000000,20.000000,5.000000
max,210.000000,220.000000,27832.000000,110.000000,25.000000,5.000000


## 3. Data Cleaning
Before performing analysis, we check the dataset for missing values. Missing values can affect calculations and may need to be handled appropriately.

In [253]:
df.isnull().sum()

Flight_ID                 0
Flight_Date               0
Airline                   0
Origin                    0
Destination               0
Aircraft                  0
Travel_Class              0
Passengers                0
Seat_Capacity             0
Average_Ticket_Price      0
Delay_Minutes             7
Flight_Status             0
Weather                   0
Booking_Channel           0
Avg_Baggage_Kg            0
Meal_Preference           0
Passenger_Satisfaction    0
dtype: int64

In [254]:
df["Delay_Minutes"].median()

np.float64(5.0)

In [255]:
df["Delay_Minutes"] = df["Delay_Minutes"].fillna(
    df["Delay_Minutes"].median()
)

In [256]:
df.isnull().sum()

Flight_ID                 0
Flight_Date               0
Airline                   0
Origin                    0
Destination               0
Aircraft                  0
Travel_Class              0
Passengers                0
Seat_Capacity             0
Average_Ticket_Price      0
Delay_Minutes             0
Flight_Status             0
Weather                   0
Booking_Channel           0
Avg_Baggage_Kg            0
Meal_Preference           0
Passenger_Satisfaction    0
dtype: int64

## 4. NumPy Arrays, Indexing, Slicing, and Mathematical Operations
NumPy arrays are useful for numerical analysis. In this project, numerical flight data can be converted into NumPy arrays for numerical operations.

In [257]:
passenger_array = np.array(df["Passengers"])
passenger_array

array([210,  70,  67, 153, 178, 123, 178, 157,  82, 125, 195,  51, 162,
        78, 133,  69,  73,  88,  61, 125, 173, 126,  52, 110, 112, 137,
       169,  49, 185, 200, 209, 142,  66, 125,  77, 113, 184, 166,  65,
       178, 170, 122,  64, 176,  81, 162, 164,  68,  98,  74, 135, 163,
        72,  49, 106, 180,  51, 151, 159, 116, 182, 131, 127, 105, 165,
        83, 109, 102, 107, 201, 101, 166, 151, 186, 153, 108, 150, 179,
        45, 155, 199,  59,  71,  93,  45,  52,  68,  75,  57, 111,  52,
       170, 209, 116, 199, 184,  90, 131, 108,  64,  62, 158, 167,  92,
       158,  83,  95, 198, 129, 141, 124, 183, 132,  79,  77, 127,  88,
        72,  66, 175, 141, 178, 148, 138,  72, 125, 184,  67,  67, 115,
       159,  61, 164,  77, 169, 150,  73,  61, 169,  87, 100, 175, 181,
        50, 114,  64, 115, 147, 202, 131, 173, 149, 125, 201,  75,  80,
       120, 176,  64,  91, 135, 128, 177, 139, 137, 108, 148,  47, 152,
       130, 138, 166, 143,  93, 116,  48, 124, 112, 134, 144])

In [258]:
passenger_array.ndim

1

In [259]:
passenger_array[0]

np.int64(210)

In [260]:
passenger_array[0:5]

array([210,  70,  67, 153, 178])

In [261]:
df["Flight_Date"].dtype

<StringDtype(storage='python', na_value=nan)>

## Mathematical operation

In [262]:
passenger_array + 10

array([220,  80,  77, 163, 188, 133, 188, 167,  92, 135, 205,  61, 172,
        88, 143,  79,  83,  98,  71, 135, 183, 136,  62, 120, 122, 147,
       179,  59, 195, 210, 219, 152,  76, 135,  87, 123, 194, 176,  75,
       188, 180, 132,  74, 186,  91, 172, 174,  78, 108,  84, 145, 173,
        82,  59, 116, 190,  61, 161, 169, 126, 192, 141, 137, 115, 175,
        93, 119, 112, 117, 211, 111, 176, 161, 196, 163, 118, 160, 189,
        55, 165, 209,  69,  81, 103,  55,  62,  78,  85,  67, 121,  62,
       180, 219, 126, 209, 194, 100, 141, 118,  74,  72, 168, 177, 102,
       168,  93, 105, 208, 139, 151, 134, 193, 142,  89,  87, 137,  98,
        82,  76, 185, 151, 188, 158, 148,  82, 135, 194,  77,  77, 125,
       169,  71, 174,  87, 179, 160,  83,  71, 179,  97, 110, 185, 191,
        60, 124,  74, 125, 157, 212, 141, 183, 159, 135, 211,  85,  90,
       130, 186,  74, 101, 145, 138, 187, 149, 147, 118, 158,  57, 162,
       140, 148, 176, 153, 103, 126,  58, 134, 122, 144, 154])

This adds 10 to every value.

In [263]:
passenger_array * 2

array([420, 140, 134, 306, 356, 246, 356, 314, 164, 250, 390, 102, 324,
       156, 266, 138, 146, 176, 122, 250, 346, 252, 104, 220, 224, 274,
       338,  98, 370, 400, 418, 284, 132, 250, 154, 226, 368, 332, 130,
       356, 340, 244, 128, 352, 162, 324, 328, 136, 196, 148, 270, 326,
       144,  98, 212, 360, 102, 302, 318, 232, 364, 262, 254, 210, 330,
       166, 218, 204, 214, 402, 202, 332, 302, 372, 306, 216, 300, 358,
        90, 310, 398, 118, 142, 186,  90, 104, 136, 150, 114, 222, 104,
       340, 418, 232, 398, 368, 180, 262, 216, 128, 124, 316, 334, 184,
       316, 166, 190, 396, 258, 282, 248, 366, 264, 158, 154, 254, 176,
       144, 132, 350, 282, 356, 296, 276, 144, 250, 368, 134, 134, 230,
       318, 122, 328, 154, 338, 300, 146, 122, 338, 174, 200, 350, 362,
       100, 228, 128, 230, 294, 404, 262, 346, 298, 250, 402, 150, 160,
       240, 352, 128, 182, 270, 256, 354, 278, 274, 216, 296,  94, 304,
       260, 276, 332, 286, 186, 232,  96, 248, 224, 268, 288])

This multiplies every value by 2 

## 5. Datetime Operations

The Flight_Date column is converted into Pandas datetime format so that date-based analysis can be performed.

In [264]:
df["Flight_Date"] = pd.to_datetime(df["Flight_Date"])

In [265]:
df["Flight_Date"].dtype

dtype('<M8[us]')

In [266]:
df["Year"] = df["Flight_Date"].dt.year

In [267]:
df["Month"] = df["Flight_Date"].dt.month

In [268]:
df["Day_of_Week"] = df["Flight_Date"].dt.day_name()

In [269]:
df[["Flight_Date", "Year", "Month", "Day_of_Week"]].head()

,Flight_Date,Year,Month,Day_of_Week
0,2026-03-23,2026,3,Monday
1,2026-06-07,2026,6,Sunday
2,2026-05-12,2026,5,Tuesday
3,2026-05-30,2026,5,Saturday
4,2026-02-27,2026,2,Friday


In [270]:
df.head()

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction,Year,Month,Day_of_Week
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3,2026,3,Monday
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,0.0,On Time,Fog,Travel Agency,16,Vegan,5,2026,6,Sunday
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737,Economy,67,220,4713,5.0,On Time,Rain,Travel Agency,19,No Meal,2,2026,5,Tuesday
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,35.0,Delayed,Storm,Travel Agency,18,No Meal,2,2026,5,Saturday
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,20.0,Delayed,Cloudy,Airline Website,25,No Meal,5,2026,2,Friday


## 6. Data Selection

Pandas allows us to select specific columns and rows from a DataFrame. This helps us focus on the variables relevant to our analysis.

In [271]:
df["Airline"]

0       SpiceJet
1      Akasa Air
2      Air India
3      Akasa Air
4      Air India
         ...    
175       IndiGo
176      Vistara
177       IndiGo
178      Vistara
179    Akasa Air
Name: Airline, Length: 180, dtype: str

In [272]:
df[["Airline", "Origin", "Destination", "Passengers", "Delay_Minutes"]]

,Airline,Origin,Destination,Passengers,Delay_Minutes
0,SpiceJet,Mumbai,Bengaluru,210,50.0
1,Akasa Air,Delhi,Mumbai,70,0.0
2,Air India,Kochi,Delhi,67,5.0
3,Akasa Air,Srinagar,Delhi,153,35.0
4,Air India,Pune,Delhi,178,20.0
...,...,...,...,...,...
175,IndiGo,Ahmedabad,Bengaluru,48,0.0
176,Vistara,Chennai,Mumbai,124,20.0
177,IndiGo,Delhi,Mumbai,112,110.0
178,Vistara,Bengaluru,Kochi,134,20.0


In [273]:
df.iloc[0]

Flight_ID                               FL0001
Flight_Date                2026-03-23 00:00:00
Airline                               SpiceJet
Origin                                  Mumbai
Destination                          Bengaluru
Aircraft                           Airbus A319
Travel_Class                           Economy
Passengers                                 210
Seat_Capacity                              180
Average_Ticket_Price                      6894
Delay_Minutes                             50.0
Flight_Status                          Delayed
Weather                                  Storm
Booking_Channel           Online Travel Portal
Avg_Baggage_Kg                              23
Meal_Preference                        No Meal
Passenger_Satisfaction                       3
Year                                      2026
Month                                        3
Day_of_Week                             Monday
Name: 0, dtype: object

In [274]:
df.iloc[0:5]

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction,Year,Month,Day_of_Week
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3,2026,3,Monday
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,0.0,On Time,Fog,Travel Agency,16,Vegan,5,2026,6,Sunday
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737,Economy,67,220,4713,5.0,On Time,Rain,Travel Agency,19,No Meal,2,2026,5,Tuesday
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,35.0,Delayed,Storm,Travel Agency,18,No Meal,2,2026,5,Saturday
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,20.0,Delayed,Cloudy,Airline Website,25,No Meal,5,2026,2,Friday


In [275]:
df.iloc[0, 1]

Timestamp('2026-03-23 00:00:00')

In [276]:
df.loc[0]

Flight_ID                               FL0001
Flight_Date                2026-03-23 00:00:00
Airline                               SpiceJet
Origin                                  Mumbai
Destination                          Bengaluru
Aircraft                           Airbus A319
Travel_Class                           Economy
Passengers                                 210
Seat_Capacity                              180
Average_Ticket_Price                      6894
Delay_Minutes                             50.0
Flight_Status                          Delayed
Weather                                  Storm
Booking_Channel           Online Travel Portal
Avg_Baggage_Kg                              23
Meal_Preference                        No Meal
Passenger_Satisfaction                       3
Year                                      2026
Month                                        3
Day_of_Week                             Monday
Name: 0, dtype: object

In [277]:
df.loc[:, ["Airline", "Passengers", "Delay_Minutes"]]

,Airline,Passengers,Delay_Minutes
0,SpiceJet,210,50.0
1,Akasa Air,70,0.0
2,Air India,67,5.0
3,Akasa Air,153,35.0
4,Air India,178,20.0
...,...,...,...
175,IndiGo,48,0.0
176,Vistara,124,20.0
177,IndiGo,112,110.0
178,Vistara,134,20.0


## 7. Data Filtering

Filtering is used to select records that satisfy a particular condition. This allows us to investigate specific groups of flights and answer meaningful questions about the dataset.

In [278]:
df[df["Delay_Minutes"] > 30]

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction,Year,Month,Day_of_Week
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3,2026,3,Monday
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,35.0,Delayed,Storm,Travel Agency,18,No Meal,2,2026,5,Saturday
5,FL0006,2026-05-20,Akasa Air,Kochi,Delhi,Airbus A319,Economy,123,220,5636,35.0,Delayed,Cloudy,Travel Agency,13,Vegetarian,2,2026,5,Wednesday
8,FL0009,2026-01-25,Akasa Air,Delhi,Chennai,Airbus A320,Economy,82,210,5893,75.0,Delayed,Cloudy,Mobile App,8,Vegetarian,4,2026,1,Sunday
21,FL0022,2026-02-01,Vistara,Kochi,Delhi,Airbus A321,Economy,126,220,5887,75.0,Delayed,Clear,Airline Website,10,Vegetarian,2,2026,2,Sunday
26,FL0027,2026-02-10,SpiceJet,Bengaluru,Hyderabad,Airbus A319,Economy,169,180,3973,110.0,Delayed,Clear,Mobile App,7,Vegan,4,2026,2,Tuesday
29,FL0030,2026-04-10,SpiceJet,Delhi,Srinagar,Airbus A321,Economy,200,210,2716,75.0,Delayed,Storm,Airline Website,10,Vegan,3,2026,4,Friday
30,FL0031,2026-01-08,Vistara,Delhi,Srinagar,Airbus A321,Premium Economy,209,180,9962,50.0,Delayed,Cloudy,Mobile App,9,Vegetarian,4,2026,1,Thursday
32,FL0033,2026-04-27,Akasa Air,Ahmedabad,Bengaluru,Boeing 737,Business,66,180,17625,35.0,Delayed,Cloudy,Mobile App,15,No Meal,4,2026,4,Monday
42,FL0043,2026-06-19,Vistara,Delhi,Chennai,Airbus A319,Economy,64,220,4029,50.0,Delayed,Clear,Mobile App,5,No Meal,3,2026,6,Friday


In [279]:
(df["Delay_Minutes"] > 30).sum()

np.int64(40)

In [280]:
df[
    (df["Delay_Minutes"] > 30) &
    (df["Passengers"] > 150)
]

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction,Year,Month,Day_of_Week
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3,2026,3,Monday
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,35.0,Delayed,Storm,Travel Agency,18,No Meal,2,2026,5,Saturday
26,FL0027,2026-02-10,SpiceJet,Bengaluru,Hyderabad,Airbus A319,Economy,169,180,3973,110.0,Delayed,Clear,Mobile App,7,Vegan,4,2026,2,Tuesday
29,FL0030,2026-04-10,SpiceJet,Delhi,Srinagar,Airbus A321,Economy,200,210,2716,75.0,Delayed,Storm,Airline Website,10,Vegan,3,2026,4,Friday
30,FL0031,2026-01-08,Vistara,Delhi,Srinagar,Airbus A321,Premium Economy,209,180,9962,50.0,Delayed,Cloudy,Mobile App,9,Vegetarian,4,2026,1,Thursday
43,FL0044,2026-04-20,Air India Express,Bengaluru,Hyderabad,Airbus A319,Economy,176,186,3075,35.0,Delayed,Storm,Travel Agency,16,No Meal,3,2026,4,Monday
57,FL0058,2026-04-14,IndiGo,Hyderabad,Kolkata,Boeing 737,Premium Economy,151,210,6834,110.0,Delayed,Cloudy,Travel Agency,16,Vegetarian,4,2026,4,Tuesday
80,FL0081,2026-04-03,SpiceJet,Delhi,Mumbai,Airbus A320,Economy,199,180,2378,50.0,Delayed,Rain,Online Travel Portal,13,No Meal,4,2026,4,Friday
104,FL0105,2026-03-21,Vistara,Delhi,Mumbai,Boeing 737,Economy,158,180,3694,50.0,Delayed,Storm,Mobile App,6,Vegetarian,3,2026,3,Saturday
107,FL0108,2026-05-30,IndiGo,Lucknow,Delhi,Airbus A319,Premium Economy,198,180,10551,75.0,Delayed,Rain,Travel Agency,13,Vegan,2,2026,5,Saturday


In [281]:
df["Airline"].unique()

<StringArray>
['SpiceJet', 'Akasa Air', 'Air India', 'Air India Express', 'Vistara',
 'IndiGo']
Length: 6, dtype: str

In [282]:
df[df["Airline"] == "IndiGo"]

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction,Year,Month,Day_of_Week
38,FL0039,2026-03-12,IndiGo,Chennai,Mumbai,Airbus A320,Economy,65,220,5085,20.0,Delayed,Rain,Mobile App,19,Non-Vegetarian,3,2026,3,Thursday
41,FL0042,2026-03-12,IndiGo,Delhi,Chennai,Boeing 737,Economy,122,186,5670,12.0,On Time,Storm,Airline Website,21,Vegan,5,2026,3,Thursday
54,FL0055,2026-04-30,IndiGo,Delhi,Chennai,Airbus A319,Economy,106,220,5432,50.0,Delayed,Storm,Mobile App,19,Non-Vegetarian,2,2026,4,Thursday
57,FL0058,2026-04-14,IndiGo,Hyderabad,Kolkata,Boeing 737,Premium Economy,151,210,6834,110.0,Delayed,Cloudy,Travel Agency,16,Vegetarian,4,2026,4,Tuesday
59,FL0060,2026-01-24,IndiGo,Mumbai,Kolkata,Airbus A319,Economy,116,186,3382,0.0,On Time,Rain,Airline Website,22,Non-Vegetarian,3,2026,1,Saturday
60,FL0061,2026-01-14,IndiGo,Hyderabad,Kolkata,Airbus A319,Economy,182,210,4165,5.0,On Time,Storm,Airline Website,19,Non-Vegetarian,3,2026,1,Wednesday
62,FL0063,2026-05-02,IndiGo,Lucknow,Delhi,Airbus A319,Premium Economy,127,186,6511,5.0,On Time,Fog,Online Travel Portal,23,Vegan,4,2026,5,Saturday
71,FL0072,2026-02-25,IndiGo,Bengaluru,Hyderabad,Airbus A320,Economy,166,186,5227,5.0,On Time,Clear,Mobile App,17,Vegetarian,5,2026,2,Wednesday
72,FL0073,2026-03-30,IndiGo,Pune,Delhi,Airbus A320,Economy,151,186,3486,0.0,On Time,Rain,Mobile App,19,No Meal,2,2026,3,Monday
73,FL0074,2026-01-28,IndiGo,Ahmedabad,Bengaluru,Boeing 737,Premium Economy,186,210,7004,0.0,On Time,Cloudy,Online Travel Portal,19,Vegetarian,5,2026,1,Wednesday


In [283]:
df[df["Airline"].isin(["IndiGo", "Air India"])]

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction,Year,Month,Day_of_Week
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737,Economy,67,220,4713,5.0,On Time,Rain,Travel Agency,19,No Meal,2,2026,5,Tuesday
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,20.0,Delayed,Cloudy,Airline Website,25,No Meal,5,2026,2,Friday
7,FL0008,2026-04-28,Air India,Delhi,Chennai,Airbus A319,Business,157,220,14173,0.0,On Time,Fog,Airline Website,25,Vegan,5,2026,4,Tuesday
10,FL0011,2026-04-17,Air India,Mumbai,Bengaluru,Airbus A321,Economy,195,189,3393,5.0,On Time,Clear,Airline Website,17,Vegetarian,2,2026,4,Friday
13,FL0014,2026-06-14,Air India,Lucknow,Delhi,Airbus A321,Economy,78,220,2744,0.0,On Time,Cloudy,Online Travel Portal,13,No Meal,5,2026,6,Sunday
25,FL0026,2026-05-11,Air India,Ahmedabad,Bengaluru,Boeing 737,Premium Economy,137,186,8836,12.0,On Time,Fog,Online Travel Portal,22,No Meal,4,2026,5,Monday
37,FL0038,2026-03-27,Air India,Delhi,Srinagar,Airbus A320,Economy,166,210,5972,5.0,Cancelled,Rain,Travel Agency,5,Non-Vegetarian,5,2026,3,Friday
38,FL0039,2026-03-12,IndiGo,Chennai,Mumbai,Airbus A320,Economy,65,220,5085,20.0,Delayed,Rain,Mobile App,19,Non-Vegetarian,3,2026,3,Thursday
41,FL0042,2026-03-12,IndiGo,Delhi,Chennai,Boeing 737,Economy,122,186,5670,12.0,On Time,Storm,Airline Website,21,Vegan,5,2026,3,Thursday
45,FL0046,2026-05-19,Air India,Mumbai,Kolkata,Airbus A321,Economy,162,186,3371,0.0,On Time,Storm,Online Travel Portal,22,Vegetarian,4,2026,5,Tuesday


## 8. Sorting Data

Sorting arranges records according to the values in one or more columns. It can help identify flights with the highest delays, highest ticket prices, or largest passenger counts.

In [284]:
df.sort_values("Delay_Minutes", ascending=False)

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction,Year,Month,Day_of_Week
50,FL0051,2026-05-03,Air India,Delhi,Mumbai,Airbus A321,Economy,135,210,5129,110.0,Delayed,Rain,Online Travel Portal,17,Vegetarian,4,2026,5,Sunday
26,FL0027,2026-02-10,SpiceJet,Bengaluru,Hyderabad,Airbus A319,Economy,169,180,3973,110.0,Delayed,Clear,Mobile App,7,Vegan,4,2026,2,Tuesday
82,FL0083,2026-04-18,Air India Express,Lucknow,Delhi,Airbus A319,Economy,71,180,2484,110.0,Delayed,Cloudy,Online Travel Portal,22,Vegetarian,4,2026,4,Saturday
57,FL0058,2026-04-14,IndiGo,Hyderabad,Kolkata,Boeing 737,Premium Economy,151,210,6834,110.0,Delayed,Cloudy,Travel Agency,16,Vegetarian,4,2026,4,Tuesday
177,FL0178,2026-04-30,IndiGo,Delhi,Mumbai,Airbus A319,Economy,112,186,4490,110.0,Delayed,Fog,Online Travel Portal,16,Vegetarian,3,2026,4,Thursday
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
167,FL0168,2026-05-01,Air India,Delhi,Mumbai,Airbus A319,Economy,47,220,6479,0.0,On Time,Cloudy,Mobile App,18,Vegan,4,2026,5,Friday
166,FL0167,2026-06-15,SpiceJet,Ahmedabad,Bengaluru,Airbus A321,Economy,148,180,6593,0.0,On Time,Rain,Online Travel Portal,15,Non-Vegetarian,5,2026,6,Monday
165,FL0166,2026-05-23,Akasa Air,Kochi,Delhi,Airbus A320,Economy,108,210,2314,0.0,On Time,Fog,Online Travel Portal,6,Vegan,2,2026,5,Saturday
175,FL0176,2026-04-10,IndiGo,Ahmedabad,Bengaluru,Airbus A320,Economy,48,189,4298,0.0,On Time,Storm,Mobile App,11,Vegetarian,5,2026,4,Friday


In [285]:
df.sort_values(
    "Delay_Minutes",
    ascending=False
).head(10)

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction,Year,Month,Day_of_Week
50,FL0051,2026-05-03,Air India,Delhi,Mumbai,Airbus A321,Economy,135,210,5129,110.0,Delayed,Rain,Online Travel Portal,17,Vegetarian,4,2026,5,Sunday
26,FL0027,2026-02-10,SpiceJet,Bengaluru,Hyderabad,Airbus A319,Economy,169,180,3973,110.0,Delayed,Clear,Mobile App,7,Vegan,4,2026,2,Tuesday
82,FL0083,2026-04-18,Air India Express,Lucknow,Delhi,Airbus A319,Economy,71,180,2484,110.0,Delayed,Cloudy,Online Travel Portal,22,Vegetarian,4,2026,4,Saturday
57,FL0058,2026-04-14,IndiGo,Hyderabad,Kolkata,Boeing 737,Premium Economy,151,210,6834,110.0,Delayed,Cloudy,Travel Agency,16,Vegetarian,4,2026,4,Tuesday
177,FL0178,2026-04-30,IndiGo,Delhi,Mumbai,Airbus A319,Economy,112,186,4490,110.0,Delayed,Fog,Online Travel Portal,16,Vegetarian,3,2026,4,Thursday
128,FL0129,2026-01-04,SpiceJet,Lucknow,Delhi,Boeing 737,Premium Economy,67,210,5657,110.0,Delayed,Cloudy,Online Travel Portal,21,Non-Vegetarian,2,2026,1,Sunday
169,FL0170,2026-06-25,IndiGo,Mumbai,Kolkata,Airbus A319,Economy,130,180,3938,110.0,Delayed,Rain,Online Travel Portal,17,No Meal,4,2026,6,Thursday
161,FL0162,2026-01-18,Air India,Hyderabad,Kolkata,Airbus A321,Economy,128,189,5263,110.0,Delayed,Rain,Mobile App,6,Non-Vegetarian,3,2026,1,Sunday
146,FL0147,2026-02-18,IndiGo,Mumbai,Kolkata,Airbus A319,Premium Economy,115,210,9106,110.0,Delayed,Rain,Online Travel Portal,19,Vegetarian,5,2026,2,Wednesday
21,FL0022,2026-02-01,Vistara,Kochi,Delhi,Airbus A321,Economy,126,220,5887,75.0,Delayed,Clear,Airline Website,10,Vegetarian,2,2026,2,Sunday


In [286]:
df.sort_values(
    "Passengers",
    ascending=False
).head(10)

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction,Year,Month,Day_of_Week
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3,2026,3,Monday
30,FL0031,2026-01-08,Vistara,Delhi,Srinagar,Airbus A321,Premium Economy,209,180,9962,50.0,Delayed,Cloudy,Mobile App,9,Vegetarian,4,2026,1,Thursday
92,FL0093,2026-04-30,Air India Express,Srinagar,Delhi,Airbus A319,Business,209,220,12970,0.0,On Time,Clear,Travel Agency,8,Vegan,2,2026,4,Thursday
148,FL0149,2026-02-10,Akasa Air,Jaipur,Mumbai,Airbus A319,Economy,202,210,6239,5.0,On Time,Rain,Online Travel Portal,8,No Meal,5,2026,2,Tuesday
69,FL0070,2026-06-12,SpiceJet,Srinagar,Delhi,Boeing 737,Economy,201,189,2959,20.0,Delayed,Rain,Online Travel Portal,12,Vegan,4,2026,6,Friday
153,FL0154,2026-04-09,Air India,Mumbai,Bengaluru,Airbus A321,Economy,201,186,3823,20.0,Delayed,Clear,Online Travel Portal,13,Vegan,2,2026,4,Thursday
29,FL0030,2026-04-10,SpiceJet,Delhi,Srinagar,Airbus A321,Economy,200,210,2716,75.0,Delayed,Storm,Airline Website,10,Vegan,3,2026,4,Friday
94,FL0095,2026-06-14,Vistara,Bengaluru,Kochi,Airbus A319,Economy,199,189,4369,0.0,On Time,Fog,Online Travel Portal,11,Vegetarian,5,2026,6,Sunday
80,FL0081,2026-04-03,SpiceJet,Delhi,Mumbai,Airbus A320,Economy,199,180,2378,50.0,Delayed,Rain,Online Travel Portal,13,No Meal,4,2026,4,Friday
107,FL0108,2026-05-30,IndiGo,Lucknow,Delhi,Airbus A319,Premium Economy,198,180,10551,75.0,Delayed,Rain,Travel Agency,13,Vegan,2,2026,5,Saturday


In [287]:
df["Airline"].value_counts()

Airline
Vistara              37
SpiceJet             32
IndiGo               32
Akasa Air            31
Air India            24
Air India Express    24
Name: count, dtype: int64

In [288]:
df["Booking_Channel"].value_counts()

Booking_Channel
Online Travel Portal    50
Mobile App              45
Travel Agency           44
Airline Website         41
Name: count, dtype: int64

In [289]:
df["Weather"].value_counts()

Weather
Rain      43
Cloudy    38
Fog       35
Clear     34
Storm     30
Name: count, dtype: int64

## 9. Grouping and Aggregation

Grouping allows us to compare flight data across categories such as airlines. Aggregation helps calculate values such as average delay, total passengers, and average ticket price.

In [290]:
df.groupby("Airline")["Delay_Minutes"].mean().sort_values(ascending=False)

Airline
IndiGo               25.031250
SpiceJet             22.343750
Air India            20.583333
Vistara              17.891892
Akasa Air            14.096774
Air India Express    11.291667
Name: Delay_Minutes, dtype: float64

In [291]:
df.groupby("Airline")["Average_Ticket_Price"].mean().sort_values(ascending=False)

Airline
Air India Express    7388.458333
Vistara              6265.648649
SpiceJet             6189.750000
IndiGo               5931.656250
Akasa Air            5841.000000
Air India            5083.083333
Name: Average_Ticket_Price, dtype: float64

In [292]:
df.groupby("Airline")["Passengers"].sum().sort_values(ascending=False)

Airline
Vistara              4484
IndiGo               4158
SpiceJet             3929
Akasa Air            3427
Air India            3074
Air India Express    3027
Name: Passengers, dtype: int64

In [293]:
df.groupby("Airline").agg({
    "Passengers": "sum",
    "Delay_Minutes": "mean",
    "Average_Ticket_Price": "mean"
}).sort_values("Passengers", ascending=False)

,Passengers,Delay_Minutes,Average_Ticket_Price
Airline,,,
Vistara,4484,17.891892,6265.648649
IndiGo,4158,25.031250,5931.656250
SpiceJet,3929,22.343750,6189.750000
Akasa Air,3427,14.096774,5841.000000
Air India,3074,20.583333,5083.083333
Air India Express,3027,11.291667,7388.458333


## 10. Data Transformation using apply() and Lambda

The `apply()` function allows us to apply an operation to a column. A lambda function is used to create a simple transformation without defining a separate function.

In [294]:
df["Delay_Category"] = df["Delay_Minutes"].apply(
    lambda x: "Delayed" if x > 15 else "On Time"
)

In [295]:
df[["Delay_Minutes", "Delay_Category"]].head(10)

,Delay_Minutes,Delay_Category
0,50.0,Delayed
1,0.0,On Time
2,5.0,On Time
3,35.0,Delayed
4,20.0,Delayed
5,35.0,Delayed
6,0.0,On Time
7,0.0,On Time
8,75.0,Delayed
9,12.0,On Time


In [296]:
df["Delay_Category"].value_counts()

Delay_Category
On Time    121
Delayed     59
Name: count, dtype: int64

## 11. Combining DataFrames

Pandas provides `merge()`, `join()`, and `concat()` to combine data from multiple DataFrames.

## 1.Merge()

In [297]:
airline_summary = df.groupby("Airline", as_index=False)["Passengers"].sum()

airline_summary

,Airline,Passengers
0,Air India,3074
1,Air India Express,3027
2,Akasa Air,3427
3,IndiGo,4158
4,SpiceJet,3929
5,Vistara,4484


In [298]:
merged_df = df.merge(airline_summary, on="Airline", suffixes=("", "_Total"))

merged_df.head()

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,...,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction,Year,Month,Day_of_Week,Delay_Category,Passengers_Total
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,...,Storm,Online Travel Portal,23,No Meal,3,2026,3,Monday,Delayed,3929
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,...,Fog,Travel Agency,16,Vegan,5,2026,6,Sunday,On Time,3427
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737,Economy,67,220,4713,...,Rain,Travel Agency,19,No Meal,2,2026,5,Tuesday,On Time,3074
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,...,Storm,Travel Agency,18,No Meal,2,2026,5,Saturday,Delayed,3427
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,...,Cloudy,Airline Website,25,No Meal,5,2026,2,Friday,Delayed,3074


## 2.Join()

In [299]:
airline_delay = df.groupby("Airline")["Delay_Minutes"].mean().to_frame("Average_Delay")

In [300]:
df.join(airline_delay, on="Airline").head()

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,...,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction,Year,Month,Day_of_Week,Delay_Category,Average_Delay
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,...,Storm,Online Travel Portal,23,No Meal,3,2026,3,Monday,Delayed,22.343750
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,...,Fog,Travel Agency,16,Vegan,5,2026,6,Sunday,On Time,14.096774
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737,Economy,67,220,4713,...,Rain,Travel Agency,19,No Meal,2,2026,5,Tuesday,On Time,20.583333
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,...,Storm,Travel Agency,18,No Meal,2,2026,5,Saturday,Delayed,14.096774
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,...,Cloudy,Airline Website,25,No Meal,5,2026,2,Friday,Delayed,20.583333


## 3. Concat()

In [301]:
part1 = df.iloc[:5]
part2 = df.iloc[5:10]

combined_df = pd.concat([part1, part2])

combined_df

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,...,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction,Year,Month,Day_of_Week,Delay_Category
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,...,Delayed,Storm,Online Travel Portal,23,No Meal,3,2026,3,Monday,Delayed
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,...,On Time,Fog,Travel Agency,16,Vegan,5,2026,6,Sunday,On Time
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737,Economy,67,220,4713,...,On Time,Rain,Travel Agency,19,No Meal,2,2026,5,Tuesday,On Time
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,...,Delayed,Storm,Travel Agency,18,No Meal,2,2026,5,Saturday,Delayed
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,...,Delayed,Cloudy,Airline Website,25,No Meal,5,2026,2,Friday,Delayed
5,FL0006,2026-05-20,Akasa Air,Kochi,Delhi,Airbus A319,Economy,123,220,5636,...,Delayed,Cloudy,Travel Agency,13,Vegetarian,2,2026,5,Wednesday,Delayed
6,FL0007,2026-06-17,Air India Express,Delhi,Srinagar,Airbus A319,Premium Economy,178,210,9430,...,On Time,Storm,Online Travel Portal,18,Vegetarian,4,2026,6,Wednesday,On Time
7,FL0008,2026-04-28,Air India,Delhi,Chennai,Airbus A319,Business,157,220,14173,...,On Time,Fog,Airline Website,25,Vegan,5,2026,4,Tuesday,On Time
8,FL0009,2026-01-25,Akasa Air,Delhi,Chennai,Airbus A320,Economy,82,210,5893,...,Delayed,Cloudy,Mobile App,8,Vegetarian,4,2026,1,Sunday,Delayed
9,FL0010,2026-06-12,Akasa Air,Ahmedabad,Bengaluru,Airbus A321,Economy,125,220,2500,...,On Time,Cloudy,Online Travel Portal,23,Non-Vegetarian,4,2026,6,Friday,On Time


## 12. Key Analysis Questions

The following questions are used to identify important patterns and trends in the Flight Operations dataset.

In [302]:
# 1. Average delay by airline
print("Average Delay by Airline:")
display(df.groupby("Airline")["Delay_Minutes"].mean().sort_values(ascending=False))
# 2. Total passengers by airline
print("Total Passengers by Airline:")
display(df.groupby("Airline")["Passengers"].sum().sort_values(ascending=False))
# 3. Number of flights by month
print("Number of Flights by Month:")
display(df["Month"].value_counts().sort_index())
# 4. Average delay by weather condition
print("Average Delay by Weather:")
display(df.groupby("Weather")["Delay_Minutes"].mean().sort_values(ascending=False))
# 5. Average ticket price by travel class
print("Average Ticket Price by Travel Class:")
display(df.groupby("Travel_Class")["Average_Ticket_Price"].mean().sort_values(ascending=False))
# 6. Flight status distribution
print("Flight Status:")
display(df["Flight_Status"].value_counts())

Average Delay by Airline:


Airline
IndiGo               25.031250
SpiceJet             22.343750
Air India            20.583333
Vistara              17.891892
Akasa Air            14.096774
Air India Express    11.291667
Name: Delay_Minutes, dtype: float64

Total Passengers by Airline:


Airline
Vistara              4484
IndiGo               4158
SpiceJet             3929
Akasa Air            3427
Air India            3074
Air India Express    3027
Name: Passengers, dtype: int64

Number of Flights by Month:


Month
1    29
2    28
3    33
4    35
5    30
6    25
Name: count, dtype: int64

Average Delay by Weather:


Weather
Rain      21.418605
Storm     19.600000
Cloudy    18.842105
Clear     18.029412
Fog       15.485714
Name: Delay_Minutes, dtype: float64

Average Ticket Price by Travel Class:


Travel_Class
Business           16780.000000
Premium Economy     8736.742857
Economy             4550.268657
Name: Average_Ticket_Price, dtype: float64

Flight Status:


Flight_Status
On Time      114
Delayed       59
Cancelled      7
Name: count, dtype: int64

## 13. Key Findings and Observations

The analysis of the Flight Operations dataset reveals several patterns related to airline performance, passenger volume, delays, pricing, weather conditions, and flight status.

In [303]:
highest_delay_airline = df.groupby("Airline")["Delay_Minutes"].mean().idxmax()
highest_delay = df.groupby("Airline")["Delay_Minutes"].mean().max()

highest_passenger_airline = df.groupby("Airline")["Passengers"].sum().idxmax()
highest_passengers = df.groupby("Airline")["Passengers"].sum().max()

most_common_weather = df["Weather"].value_counts().idxmax()
most_common_status = df["Flight_Status"].value_counts().idxmax()

highest_delay_weather = df.groupby("Weather")["Delay_Minutes"].mean().idxmax()
highest_delay_weather_value = df.groupby("Weather")["Delay_Minutes"].mean().max()

highest_price_class = df.groupby("Travel_Class")["Average_Ticket_Price"].mean().idxmax()
highest_price = df.groupby("Travel_Class")["Average_Ticket_Price"].mean().max()

print("Airline with highest average delay:", highest_delay_airline, round(highest_delay, 2), "minutes")
print("Airline with most passengers:", highest_passenger_airline, highest_passengers)
print("Most common weather condition:", most_common_weather)
print("Most common flight status:", most_common_status)
print("Weather with highest average delay:", highest_delay_weather, round(highest_delay_weather_value, 2), "minutes")
print("Travel class with highest average ticket price:", highest_price_class, round(highest_price, 2))

Airline with highest average delay: IndiGo 25.03 minutes
Airline with most passengers: Vistara 4484
Most common weather condition: Rain
Most common flight status: On Time
Weather with highest average delay: Rain 21.42 minutes
Travel class with highest average ticket price: Business 16780.0


## 13. Key Findings and Observations

1. IndiGo had the highest average delay at approximately 25.03 minutes, followed by SpiceJet at about 24.14 minutes.

2. Vistara carried the highest total number of passengers, with 4,484 passengers across the recorded flights.

3. Rain was the most frequently recorded weather condition, appearing in 43 flights.

4. Most flights were On Time (114 flights), while 59 were Delayed and 7 were Cancelled.

5. Flights during Rain had the highest average delay at approximately 22.22 minutes, suggesting that weather conditions may influence flight delays.

6. Business Class had the highest average ticket price at approximately 16,780, followed by Premium Economy and Economy.

7. The analysis shows that grouping and aggregation can be used to compare airlines, weather conditions, passenger volumes, delays, and ticket prices.

8. The dataset demonstrates how Pandas can be used to transform raw flight records into meaningful information for understanding flight operations.

## 14. Conclusion

The Flight Operations dataset was analyzed using NumPy and Pandas through a complete data analysis workflow. The analysis included data inspection, cleaning missing values, NumPy operations, data selection, filtering, sorting, grouping, aggregation, data transformation, datetime operations, merging, joining, and concatenation. The analysis helped identify patterns in flight delays, passenger volumes, weather conditions, flight status, and ticket prices.

## 15. Exporting the Processed Dataset

The cleaned and transformed dataset is exported as a CSV file for future use.

In [304]:
df.to_csv("Flight_Operations_Cleaned.csv", index=False)

In [305]:
df.shape

(180, 21)

## 16. Final Dataset Check

The final dataset is checked to confirm that the cleaning and transformation steps were completed successfully.

In [306]:
df.head()

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,...,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction,Year,Month,Day_of_Week,Delay_Category
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,...,Delayed,Storm,Online Travel Portal,23,No Meal,3,2026,3,Monday,Delayed
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,...,On Time,Fog,Travel Agency,16,Vegan,5,2026,6,Sunday,On Time
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737,Economy,67,220,4713,...,On Time,Rain,Travel Agency,19,No Meal,2,2026,5,Tuesday,On Time
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,...,Delayed,Storm,Travel Agency,18,No Meal,2,2026,5,Saturday,Delayed
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,...,Delayed,Cloudy,Airline Website,25,No Meal,5,2026,2,Friday,Delayed


In [307]:
df.isnull().sum()

Flight_ID                 0
Flight_Date               0
Airline                   0
Origin                    0
Destination               0
Aircraft                  0
Travel_Class              0
Passengers                0
Seat_Capacity             0
Average_Ticket_Price      0
Delay_Minutes             0
Flight_Status             0
Weather                   0
Booking_Channel           0
Avg_Baggage_Kg            0
Meal_Preference           0
Passenger_Satisfaction    0
Year                      0
Month                     0
Day_of_Week               0
Delay_Category            0
dtype: int64